ArmPkg/Library/ArmFfaLib/ArmFfaCommon.c

```diff
@@ -0,0 +1,758 @@
+/** @file
+  Arm Ffa library common code.
+
+  Copyright (c) 2024, Arm Limited. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+   @par Glossary:
+     - FF-A - Firmware Framework for Arm A-profile
+
+   @par Reference(s):
+     - Arm Firmware Framework for Arm A-Profile [https://developer.arm.com/documentation/den0077/latest]
+
+**/
+#include <Uefi.h>
+#include <Pi/PiMultiPhase.h>
+
+#include <Library/ArmLib.h>
+#include <Library/ArmFfaLib.h>
+#include <Library/ArmSmcLib.h>
+#include <Library/ArmSvcLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/PcdLib.h>
+
+#include <IndustryStandard/ArmFfaSvc.h>
+#include <IndustryStandard/ArmFfaPartInfo.h>
+
+#include "ArmFfaCommon.h"
+
+BOOLEAN  gFfaSupported;
+UINT16   gPartId;
+
+/**
+  Convert EFI_GUID to UUID format.
+  for example, If there is EFI_GUID named
+  "378daedc-f06b-4446-8314-40ab933c87a3",
+
+  EFI_GUID is saved in memory like:
+     dc ae 8d 37
+     6b f0 46 44
+     83 14 40 ab
+     93 3c 87 a3
+
+  However, UUID should be saved like:
+     37 8d ae dc
+     f0 6b 44 46
+     83 14 40 ab
+     93 3c 87 a3
+
+  FF-A and other software components (i.e. linux-kernel)
+  uses below format.
+
+  @param [in] Guid            EFI_GUID
+  @param [out] Uuid           Uuid
+
+**/
+STATIC
+EFIAPI
+VOID
+ConvertEfiGuidToUuid (

```

Comment:
No, according to your function  if changed  for "378daedc-f06b-4446-8314-40ab933c87a3",

from (EFI_GUID):
    dc ae 8d 37
    6b f0 46 44
    83 14 40 ab
    93 3c 87 a3 

to:
   dc ae 8d 37
   46 44 6b f0
   ab 40 14 83
   a3 87 3c 93

Plese see the UUID_INIT in linux kernel to know how they should be saved.
UUID should be saved like

     37 8d ae dc
     f0 6b 44 46
     83 14 40 ab
     93 3c 87 a3

Reply:
But when using the original function you had, I am getting an `INVALID_PARAMETERS` when doing `FFA_PARTITION_INFO_GET`. Does this function work with your environment? Although my SPMC is hafnium in this case.

Reply:
For reference, `0x378daedc 0xf06b4446 0x831440ab 0x933c87a3` is what I had in the dts, and `{ 0x378daedc, 0xf06b, 0x4446, { 0x83, 0x14, 0x40, 0xab, 0x93, 0x3c, 0x87, 0xa3 } }` is the EFI_GUID for standalone mm.

Reply:
Hmm, checking back on an older version of discussion: https://github.com/tianocore/edk2/pull/6116#discussion_r1774033375. I think how the UUID is populated is the issue...

Reply:
I think the manifest's uuid definiation is wrong (Sorry If you follow my former advice, that was wrong).

the manifest for StandaloneMm should be:

uuid = <0xdcae8d37 0x46446bf0 0xab401483 0xa3873c93>;

Reply:
I think we are getting in the weeds of making the code to work on both ends, which involves a lot of implementation alternations.

How about we just do not do conversion in code? In as the manifest uuid should be configured correctly to contain the correct byte order, so that the firmware/software code does not need to carry this logic.

Reply:
> How about we just do not do conversion in code? In as the manifest uuid should be configured correctly to contain the correct byte order, so that the firmware/software code does not need to carry this logic.

Hmm but pre-exist secure partition with optee defines uuid as i said.
i.e) for optee some sp, the "486178e0-e7f8-11e3-bc5e0002a5d5c51b"
the manifest defines <0xe0786148 0xe311f8e7 0x02005ebc 0x1bc5d5a5> 

to save it with 
    48 61 78 e0
     e7 f8 11  e3
    bc 5e 00 02 
    a5 d5 c5 1b

and kernel driviver probing this partition expecting save uuid with this format.

I think this horrible is because of manifest defines uuid with "uint32[4]"
and use it without any conversion (that's when writing manifest'uuid need to consider byte-order)"

and SPMC saved uuid as with above format, so whenever partition_get_info with uuid, we couldn't avoid the convertsion to meet this saved format of uuid....

Reply:
But I think the key is that we should keep the byte order the same. Would it be better to have the optee sp consumer to mangle the GUID/UUID before sending to FF-A, instead of mangling the UUID here?

At this point, the uint32[4] representation seems meaningless. But the byte order is.

I understand that partition_get_info return the uuid, but my expectation is that it will return the byte order inherited from the uuid in manifest? In this case, if we put the uuid in manifest correctly, the UEFI side will consume it as a normal GUID without any conversion just fine, right?

Reply:
Without carrying the conversion inside ff-a call, the logistic is that if this GUID is defined in UEFI, the manifest should match the UEFI definition in terms of byte order.

In your optee case, if it is optee centric and defined there, it makes better sense to have the UEFI agent that talks to this optee sp to mangle the GUID to match the byte order before sending to FF-A.

Reply:
What I say is that for memory saved value 
i.e) 378daedc-f06b-4446-8314-40ab933c87a3

EFI_GUID:
    dc ae 8d 37
    6b f0 46 44 
    83 14 40 ab
    93 3c 87 a3

other softwoare compoment saved:
    37 8d ae dc
    f0 6b 44 46
    83 14 40 ab
    93 3c 87 a3

FWIW is that except UEFI, all other agents use UUID with below.
and all other usecase follow below.
That's why i think it need to UEFI should change EFI_GUID to UUID which other agents follow

Reply:
> 37 8d ae dc
> f0 6b 44 46
> 83 14 40 ab
> 93 3c 87 a3

I think this ^ is the problem. May I know why? This is already a breaking change, what is the harm to have the manifest update to <0x378daedc 0x4446f06b 0xab401483 0xa3873c93> and not having to perform this conversion?

The fact that this GUID (and its associated byte order) is fully defined in UEFI specification, yet UEFI still needs to mangle this GUID in order to communicate to this entity behind that GUID does not make sense.

Reply:
> 37 8d ae dc
f0 6b 44 46
83 14 40 ab
93 3c 87 a3

But according to Specification RFC4122, It seems right. it should be saved as BIGENDIAN order.

> what is the harm to have the manifest update to <0x378daedc 0x4446f06b 0xab401483 0xa3873c93> and not having to perform this conversion?

Yes. there is no spmc which conversion. Just saved it as little endian uint array.
I think there should clarify how manifest should input UUID and proper conversion when it saves in memory.
But right now, it should input it to follow uuid memory saved way.

Reply:
But PI specification has defined it here:
https://uefi.org/specs/PI/1.8/V4_UEFI_Protocols.html#efi-mm-communication2-protocol

And the interpretation of the GUID is https://uefi.org/specs/UEFI/2.10/Apx_A_GUID_and_Time_Formats.html#text-representation-relationships-apxa-guid-and-time-formats.

The above should translate to the following byte order in memory, correct?
> dc ae 8d 37
> 6b f0 46 44
> 83 14 40 ab
> 93 3c 87 a3

I do not understand why we should follow the wrong byte order in memory?

Once this change is checked in, others will follow the same semantic order to populate their manifest for other services, such as TPM. Then OS will need to mangle it the same way so that they can communicate to it. Another change to fix it will be very difficult at that point.